### ISSUE TITLE
iOS: Add support to change camera resolution

### ISSUE THREAD

**Issue by g0dpain** [Redth/ZXing.Net.Maui#31]
The ability to scan linear barcodes on newer iOS devices is unreliable because the set resolution is too low.
The preset 640x480 results a lot in false values on devices like iPhone 13 Pro.

Since the iPhone 4 the minimum camera resolution is 1080p (or higher) which would work on top devices like the iPhone 13 Pro Max.
Although setting a value too great may crash the app.

Therefore a solution which enables the developer to change the resolution iOS-only would be great e.g. adding another property ``CameraResolution`` to ``BarcodeScanningOptions`` 
Another option would be to read the devices information and take the best possible value

What is your opinion @Redth ?
If you have your handsful I can provide some implementation

**Comment by Redth** [comment]
Yeah in my old Xamarin library I provided a delegate you could set to pick and return from a list of supported resolutions, the one you wanted to use....

This is a bit trickier to do in the camerax api on android, and is not exactly necessary either given that it will try and pick the closest available resolution to what you ask for ([according to the docs](https://developer.android.com/reference/androidx/camera/core/ImageAnalysis.Builder#setTargetResolution(android.util.Size))).

So, I think a nice API here would be to specify a target resolution, pass that value directly to android's API's by basically adding a new line below this one: https://github.com/Redth/ZXing.Net.Maui/blob/main/ZXing.Net.MAUI/Platforms/Android/CameraManager.android.cs#L65 `.SetTargetResolution(new Android.Util.Size(YOUR_WIDTH, YOUR_HEIGHT))` that passes in what was set on the xplat control.

For iOS I think that we'd need to introduce the logic to pick the best closest available size similar to what android does automatically.  For example right now it's just hard coded to use 640x480 here as well: https://github.com/Redth/ZXing.Net.Maui/blob/main/ZXing.Net.MAUI/Apple/CameraManager.ios.maccatalyst.cs#L39  But you can get a list of available sizes and do the math to pick the best one.

**Comment by Redth** [comment]
I started this branch https://github.com/Redth/ZXing.Net.Maui/tree/dev/target-capture-resolution

Haven't tested the android stuff yet... there's some code in there to help with the general cleanup which may fix some other lifecycle related issues on android too...

**Comment by g0dpain** [comment]
I played around a bit and found a bit of a solution..
``AVCaptureDevice`` has a function ``SupportsAVCaptureSessionPreset(preset : string)`` to determine if a passed ``AVCaptureSession.Preset`` is supported
[See Apple doc](https://developer.apple.com/documentation/avfoundation/avcapturedevice/1386263-supportssessionpreset)

```csharp
NSString gd_findBestAvCaptureSessionPreset()
{
        for (int i = availablResolutions.Count - 1; i >= 0; i--)
        {
            KeyValuePair<NSString, MSize> resolution = availablResolutions.ElementAt(i);

            if (captureDevice.SupportsAVCaptureSessionPreset(resolution.Key))
                return resolution.Key;
        }
      
        return AVCaptureSession.Preset640x480;
}
```

Although this code relies on a sorted ``availableResolutions`` dictionary.

I have tested this primarily with an iPhone 5S since every new generation supports 4k on the rear. 
Switching between front/rear worked also for me. For this device I got ``AVCaptureSession.Preset1920x1080`` for the camera on the rear and ``AVCaptureSession.Preset1280x720`` for the front camera.
[See device specs](https://support.apple.com/kb/sp738)

**Comment by g0dpain** [comment]
I'm sorry, I misunderstood what you are trying to do

In my opinion, it would be better to filter ``availablResolutions`` before looping in ``findBestAVCaptureSessionPreset(MSize target)`` to prevent crashes
[crash.txt](https://github.com/Redth/ZXing.Net.Maui/files/9469506/crash.txt)
Therefore created a PR

And wouldn't it be better to get the best possible resolution instead of 480p ?

**Comment by Redth** [comment]
Yeah good point to filter the resolutions and get a subset confirmed to be available on iOS. 

I don't think you always want the best / highest resolution so I think it's most appropriate to allow you to specify. 

What we could also do is use Size with max values as a way to indicate you want the highest possible resolution.

**Comment by g0dpain** [comment]
That would be an option too.

Let's say we try to pass 4000x4000, a resolution exceeding the given ones. The set resolution would currently still be 640x480, probably a behaviour a developer won't except or doesn't want.

After some thought I think in this particular case it is maybe enough to set the lowest possible resolution if the passed one is smaller and the highest when it is too high. That way we could pass some *ridiculous* value like 99999x99999 to ensure the highest resolution even if iOS devices are getting a better camera..

**Comment by jfversluis** [comment]
Closing this as addressed by #304, which adds `CameraResolutionSelector` and wires resolution selection through Android, iOS/MacCatalyst, and Windows.

Please try the latest build once #304 is available and let us know if you still need additional resolution-control behavior beyond the selector API.
